include/oneapi/dpl/pstl/algorithm_ranges_impl.h

```diff
@@ -544,29 +545,59 @@ __pattern_copy(__serial_tag</*IsVector*/std::false_type>, _ExecutionPolicy&&, _I
 // pattern_copy_if
 //---------------------------------------------------------------------------------------------------------------------
 
-template <typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _Pred,
+template <typename _IsVector, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _Pred,
           typename _Proj>
 std::ranges::copy_if_result<std::ranges::borrowed_iterator_t<_InRange>, std::ranges::borrowed_iterator_t<_OutRange>>
-__pattern_copy_if_ranges(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r, _Pred __pred,
-                         _Proj __proj)
+__pattern_copy_if_ranges(__parallel_tag<_IsVector> __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r,
+                         _OutRange&& __out_r, _Pred __pred, _Proj __proj)
 {
-    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+    auto __first_in = std::ranges::begin(__in_r);
+    auto __first_out = std::ranges::begin(__out_r);
+    auto __sz_in = std::ranges::size(__in_r);
+    auto __sz_out = std::ranges::size(__out_r);
 
-    oneapi::dpl::__internal::__unary_op<_Pred, _Proj> __pred_1{__pred, __proj};
+    if (__sz_in > 0 && __sz_out > 0)
```

Comment:
Does it make sense to redirect to an ordinary pattern in case of `__sz_out >= __sz_in`?

My conjecture is that  `__sz_out == __sz_in` should be common, and `__pattern_bounded_copy_if` is slower than `__pattern_copy_if` due to additional checks.